plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduPageSink.java

```diff
@@ -102,35 +101,35 @@ private KuduPageSink(
         this.generateUUID = mapping.isGenerateUUID();
 
         this.table = table;
-        this.session = clientSession.newSession();
+        this.session = clientSession;
         uuid = UUID.randomUUID().toString();
     }
 
     @Override
     public CompletableFuture<?> appendPage(Page page)
     {
-        for (int position = 0; position < page.getPositionCount(); position++) {
-            Upsert upsert = table.newUpsert();
-            PartialRow row = upsert.getRow();
-            int start = 0;
-            if (generateUUID) {
-                String id = format("%s-%08x", uuid, nextSubId++);
-                row.addString(0, id);
-                start = 1;
-            }
-
-            for (int channel = 0; channel < page.getChannelCount(); channel++) {
-                appendColumn(row, page, position, channel, channel + start);
-            }
-
-            try {
-                applyOperationAndVerifySucceeded(session, upsert);
-            }
-            catch (KuduException e) {
-                throw new RuntimeException(e);
+        try (KuduOperationApplier operationApplier = KuduOperationApplier.fromKuduClientSession(session)) {
+            for (int position = 0; position < page.getPositionCount(); position++) {
+                Upsert upsert = table.newUpsert();
+                PartialRow row = upsert.getRow();
+                int start = 0;
+                if (generateUUID) {
+                    String id = format("%s-%08x", uuid, nextSubId++);
```

Comment:
Q: Should `nextSubId` be marked `volatile` (it's a member being updated in a CompletableFuture)?

Reply:
The completablefuture `NOT_BLOCKED` is just a placeholder to signal the page sink is not blocked. `NOT_BLOCKED` does not hold a reference to `nextSubId`, so I do not think there are any issues here.

Reply:
Just to be clear, non-volatile means that any other thread may not see the new value of `nextSubId++` (it's not volatile, so there's no guarantee a separate thread will ever see the updated value, if it had ever read the value before). 
If that's not important, then feel free to ignore! (But as an aside, I think it's generally better to be explicit about these scenarios, and I would mark it volatile to be correct - rather than have code which could be "incorrect" under certain circumstances, and just depending on those situations being "impossible" - someone may always re-use this code and fall foul of the invisible update!).

Reply:
Concrete example:
```
executor.submit(() -> kuduPageSink.appendPage(page));
executor.submit(() -> kuduPageSink.appendPage(anotherPage));
```
If that were a multi-threaded Exector, this line could yield the same value in both threads:
```
String id = format("%s-%08x", uuid, nextSubId++);
```
(both threads read kuduPageSink.nextSubId , both increment it, neither see the value of the other, so it's both a race and a lost update - but even if only 1 thread did updates, because it's not volatile, there's no guarantee anyone who has a reference to kuduPageSink instance will see the update `nextSubId++`, hence the problem).

On reflection, I'd use `AtomicInteger.getAndIncrement()` for simplicity sake (and to avoid the race I mentioned - volatile wouldn't protect against that...).

But anyways - just wanted to throw that out there for your consideration!

Reply:
Correct if I'm wrong, but page sinks are not expected to be thread safe. 

No where in the javadoc for `ConnectorPageSink` does it list thread safety as a requirement. The closest thing I could find was:

```
    /**
     * Returns a future that will be completed when the page sink can accept
     * more pages.  If the page sink can accept more pages immediately,
     * this method should return {@code NOT_BLOCKED}.
     */
    CompletableFuture<?> appendPage(Page page);
``` 

Here we only return `NOT_BLOCKED` after the critical section, so there is no risk of concurrent access to `nextSubId`